File: 12-Composite/src/go/directory.go
package main

import "fmt"

// Composite — a directory that contains files and/or other directories
type Directory struct {
	Name     string
	children []FileSystemNode
}

func NewDirectory(name string) *Directory {
	return &Directory{Name: name}
}

func (d *Directory) Add(node FileSystemNode) {
	d.children = append(d.children, node)
}

func (d *Directory) GetChildren() []FileSystemNode {
	return d.children
}

func (d *Directory) GetName() string { return d.Name }

func (d *Directory) GetSize() int {
	total := 0
	for _, child := range d.children {
		total += child.GetSize()
	}
	return total
}

func (d *Directory) Print(indent string) {
	fmt.Printf("%sD %s/ (%d bytes)\n", indent, d.Name, d.GetSize())
	for _, child := range d.children {
		child.Print(indent + "  ")
	}
}

File: 12-Composite/src/java/chapter12/Directory.java
package chapter12;

import java.util.ArrayList;
import java.util.List;

/**
 * Composite — a directory that contains files and/or other directories.
 * getSize() recursively sums all children.
 */
public class Directory implements FileSystemNode {
    private final String name;
    private final List<FileSystemNode> children = new ArrayList<>();

    public Directory(String name) {
        this.name = name;
    }

    public void add(FileSystemNode node) {
        children.add(node);
    }

    public void remove(FileSystemNode node) {
        children.remove(node);
    }

    public List<FileSystemNode> getChildren() {
        return children;
    }

    @Override
    public String getName() {
        return name;
    }

    @Override
    public int getSize() {
        int total = 0;
        for (FileSystemNode child : children) {
            total += child.getSize();  // polymorphic — works for File AND Directory
        }
        return total;
    }

    @Override
    public void print(String indent) {
        System.out.println(indent + "📁 " + name + "/ (" + getSize() + " bytes)");
        for (FileSystemNode child : children) {
            child.print(indent + "  ");  // recursive — children print themselves
        }
    }
}
